### PR TITLE
WT-8703 Extend recovering from backup to allow partial restores

### PR DESCRIPTION
**Pull request by jiechenbo**
The idea behind this ticket is to provide functionality to perform a partial restore from a backup. The use case is that the user can copy over selected tables, along with passing in a target uri list. From WiredTiger's perspective, this means we must clean all the metadata entry references and history entries to the tables that are not part of the database anymore. 

Some technical key points to review:
- To cleanly remove all metadata entry references, we must perform a schema_drop on all the tables that are not in the target uri list.
- There is a hashtable performance optimisation to search the target uris quickly to an O(N) solution.
- We must generate a btree id removal list, such that recovery can truncate all history records to do with the id.
- There can only be table: formats in the target uri list.

### CODE REVIEW

File: dist/api_data.py
Comment (sueloverso) [inline dist/api_data.py:1145]: The description is no longer correct. It is not set to `true`. I suggest changing the name to be consistent with the backup target naming and call it `backup_restore_target`. The description could be something like:
If non-empty and restoring from a backup, restore only the table object targets listed. WiredTiger will remove all table entries that are not listed in the list from the reconstructed metadata. All targets must be use table: object URI configuration.

File: src/conn/conn_api.c
Comment (sueloverso) [inline src/conn/conn_api.c:2982]: This code (or the new hash code it becomes) should be in its own function. You can send in the `cfg` and check for the list in the function.

File: src/meta/meta_turtle.c
Comment (sueloverso) [inline src/meta/meta_turtle.c:149]: Since we know that mongodb table names are generally consistent and mostly fixed length, it would be better to keep track of the max length seen so far and only reallocate if the new length exceeds the already allocated one rather than allocate and free every time through here.

**Comment by jiechenbo** [comment]
Hey @sueloverso and @kommiharibabu I believe everything is ready for review now!

File: test/suite/test_backup27.py
Comment (sueloverso) [inline test/suite/test_backup27.py:81]: This comment is not up to date. This is copied from the other test that created a new table after the backup cursor was opened.

File: test/suite/test_backup27.py
Comment (sueloverso) [inline test/suite/test_backup27.py:89]: Add a comment here about testing using the history store.

**Review by sueloverso — APPROVED** [review]
LGTM. This is very close. I have a lot of minor comments and one bug fix (regarding checking for WiredTiger files). But I think all the changes are minor and I'll approve pending the minor fixes.

**Comment by sueloverso** [comment]
@agorrod I added you as a reviewer since we pulled you into this ticket that became surprisingly complicated.

File: src/meta/meta_turtle.c
Comment (sueloverso) [inline src/meta/meta_turtle.c:145]: This line should probably use a schema worker function so that it doesn't assume the naming structure. But I don't want to hold up this ticket for that. Can you create a new follow-on ticket to look investigate that?

File: src/meta/meta_turtle.c
Comment (jiechenbo) [reply src/meta/meta_turtle.c:111]: This edge case is interesting. So I took a look at the metadata entries for WiredTiger HS table, and found that it only has a `file:` entry and no table metadata entry. In our case, we only parse through table metadata entries, so I think we are fine? @sueloverso let me know your thoughts.

Comment (sueloverso) [reply src/meta/meta_turtle.c]: That would be fine. Maybe add a comment and assert that the `table:` is not in the WiredTiger namespace. That way if the history store or other WT table ever changes it will trigger and for now it will be fine.

File: src/meta/meta_turtle.c
Comment (jiechenbo) [reply src/meta/meta_turtle.c:173]: Since we provide the assert that there should be no WT table in this list, I believe we don't need to do that here anymore. Do you agree @sueloverso ?

File: src/conn/conn_api.c
Comment (agorrod) [inline src/conn/conn_api.c:1056]: The conn_api.c file is already big - could the new code mostly live in a different file?

I'd suggest using __conn_add_backup_target_uri - since I think it's better if the name clearly indicates what it's being saved for.

File: src/conn/conn_api.c
Comment (agorrod) [inline src/conn/conn_api.c:1071]: Why did you decide to wrap this in a macro? It's only used here, so I'd have thought the insert along with a comment would be adequate.

File: src/conn/conn_api.c
Comment (agorrod) [inline src/conn/conn_api.c:1107]: It's a genuine question... Is it safe to assume that there is a string in the retrieved config item?

File: src/conn/conn_api.c
Comment (agorrod) [inline src/conn/conn_api.c:1102]: The code in WT_PREFIX_MATCH relies on `strlen` on `k.str` working, but other code here passes in a `k.len` directly. I don't think it's OK to rely on `strlen`, since the retrieved config item is pointing directly to part of the total configuration string (I could be wrong, it might be that `wt_config_next` copies it out?)

File: src/meta/meta_turtle.c
Comment (agorrod) [inline src/meta/meta_turtle.c:431]: Replace "partial backup is requires" with "partial backup requires"

File: src/conn/conn_api.c
Comment (agorrod) [inline src/conn/conn_api.c:1711]: I think this is OK here, but it feels like a very general place to put the hash bucket and to always allocate one. Did you think about instead passing the `cfg` to `__wt_turtle_init` and keeping the data structures local to the backup code in there?
It looks like you need to create and save an associated ID list for RTS, and that probably needs to belong in the WT_CONNECTION (since it's shared across subsystems), but the hash table here appears more localised.

**Review by agorrod — CHANGES_REQUESTED** [review]
Thanks Jie - the shape of this change generally looks good to me, but I've got a few specific suggestions, and some questions about where the code lives.

Take a look, I'm happy to chat about any of the feedback, if it was hard to understand or you disagree with the intent.

File: dist/api_data.py
Comment (jiechenbo) [reply dist/api_data.py]: Yeah WiredTiger will remove all the contents related to the table. I have fixed up the description.

File: src/conn/conn_api.c
Comment (jiechenbo) [reply src/conn/conn_api.c]: I have placed this in metadata file now, let me know what you think.

File: src/conn/conn_api.c
Comment (jiechenbo) [reply src/conn/conn_api.c]: I have done the initialisation and destruction of the hashtable inside `__wt_turtle_init` instead now. I have maintained the backup hash in the connection structure still. The reason is because we will need to access at multiple mounts across different functions.

File: src/conn/conn_api.c
Comment (jiechenbo) [reply src/conn/conn_api.c]: This is interesting, so I grabbed the functionality of parsing the config string from statistics_log.sources inside `__statlog_config`. The code doesn't check if the string is safe or not. So from my understanding and taking that as the source of truth, I assume it is safe? My understanding is quite limited here.

File: src/conn/conn_api.c
Comment (jiechenbo) [reply src/conn/conn_api.c]: This is also interesting, since I grabbed the functionality of parsing the config string from statistics_log.sources inside `__statlog_config`.

 I have checked and printed the string of k.str and it does print out the part of the total config string. Therefore the `WT_PREFIX_MATCH` will be matching across the whole string. I can see this can be bring some performance implications. I am thinking we can create a second macro where we pass in the length as well? What are your thoughts @agorrod

File: src/meta/meta_turtle.c
Comment (sueloverso) [inline src/meta/meta_turtle.c:395]: Add a blank line between functions.

File: src/meta/meta_turtle.c
Comment (sueloverso) [inline src/meta/meta_turtle.c:416]: Now that it is in here, the hash bucket could be a local variable that gets passed into the other functions. It is good it is allocated and freed only in here. I understand the convenience of having the connection available everywhere. We are single threaded here guaranteed, and the session is also available.

Comment (agorrod) [reply src/meta/meta_turtle.c:416]: Yes - I'm with Sue, that I'd keep the bucket local and pass it as a new parameter where it's needed.

Comment (jiechenbo) [reply src/meta/meta_turtle.c]: Done!

File: src/conn/conn_api.c
Comment (jiechenbo) [reply src/conn/conn_api.c]: Talked with @agorrod and found that WT_PREFIX_MATCH does a strlen with the prefix only, which in our case is "table:".

**Comment by sueloverso** [comment]
Thanks for making the changes @jiechenbo. I know the execution team is waiting for this. @agorrod and @kommiharibabu can you review (or remove yourself from the list) and hopefully this can get merged.

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: dist/api_data.py
Comment (jiechenbo) [reply dist/api_data.py:1142]: From my understanding @kommiharibabu, I think we do require that we only do a partial restore because the current implementation does require the metadata backup file. I can see possible ways where we can extend this functionality to look at the current state of the metadata file instead, but I think that benefit should be completed in a follow-up ticket. What are your thoughts?

Comment (jiechenbo) [reply dist/api_data.py:1142]: I realised that this might be actually a cool idea we can extend @sueloverso  @agorrod what are your thoughts on this?

Comment (agorrod) [reply dist/api_data.py:1142]: I agree - it's an interesting idea, but would require more work, since the metadata manipulation would need to happen in a different place. It does make me wonder if we should update the name of the configuration option, so that we can make that change in the future.
I'd leave that one to @sueloverso, since her naming game is strong, and she understands much better how easily this would be extended to work without a backup, which probably influences how hard we should try to make the configuration option name generic.

File: src/txn/txn_rollback_to_stable.c
Comment (agorrod) [inline src/txn/txn_rollback_to_stable.c:1533]: I had to think about this a bit. I guess it's guaranteed that RTS won't touch any of the tables *not* being restored, since it doesn't visit them because they aren't in the metadata. I can also see that this step needs to be here because it's kind of like post recovery cleanup. OTOH: it would have been nice not to have this special case in RTS.
That's a general tangent. My only point is that this is a piece of logic we'll need to watch out for, and ensure it remains correct and relevant, since it's not an obvious place to look for something associated with partial restores.

**Review by agorrod — APPROVED** [review]
LGTM

File: dist/api_data.py
Comment (sueloverso) [reply dist/api_data.py:1142]: I don't think there is a future for people manually removing files from an active database and wanting to restore and run it partially. We strongly discourage them from manipulating the filespace and backup is really the only avenue to legitimately do that. 

I could see expanding the backup `target=` API or meaning to include WT files so that you get a working smaller database from the backup cursor. MongoDB cannot do that because at the "take the backup" time they don't know what they need.

If the partial restore is expanded then it will be a real project and its API can be considered at that time. Modifying the API at that point would be fine.

Comment (jiechenbo) [reply dist/api_data.py:1142]: Makes sense! It sounds like there won't be much benefit implementing this outside of backup. I am resolving this conversation

**Pull request merged**
